python/selfie-lib/selfie_lib/Snapshot.py

```diff
@@ -1,106 +1,69 @@
-        """
-        The actual implementation to add a facet. This method should only be called
-        with a SnapshotValue instance as value.
-        """
-        if not key:
-            raise ValueError("The empty string is reserved for the subject.")
-        facet_data = dict(self._facet_data)
-        facet_data[key] = value  # Here, value is guaranteed to be a SnapshotValue
-        return Snapshot(self._subject, facet_data)
+        new_facet_data = self._facet_data.plus(key, value)
+        return Snapshot(self._subject, new_facet_data)
 
     def plus_or_replace(self, key: str, value: SnapshotValue) -> "Snapshot":
         if not key:
             return Snapshot(value, self._facet_data)
-        facet_data: Dict[str, SnapshotValue] = dict(self._facet_data)
-        facet_data[self._unix_newlines(key)] = value
-        return Snapshot(self._subject, facet_data)
+        return Snapshot(self._subject, self._facet_data.plus(key, value))
 
-    def subject_or_facet_maybe(self, key: str) -> Optional[SnapshotValue]:
-        if not key:
-            return self._subject
-        return self._facet_data.get(key)
+    def subject_or_facet_maybe(self, key: str) -> Union[SnapshotValue, None]:
+        try:
+            return self._facet_data[key]
+        except KeyError:
+            return None if key else self._subject
 
     def subject_or_facet(self, key: str) -> SnapshotValue:
         value = self.subject_or_facet_maybe(key)
         if value is None:
-            raise KeyError(f"'{key}' not found in {list(self._facet_data.keys())}")
+            raise KeyError(f"'{key}' not found")
         return value
 
-    def all_entries(self) -> Dict[str, SnapshotValue]:
-        entries: Dict[str, SnapshotValue] = {"": self._subject}
-        entries.update(self._facet_data)
-        return entries
-
     @staticmethod
     def of(data: Union[bytes, str, SnapshotValue]) -> "Snapshot":
-        if isinstance(data, (bytes, str)):
-            data = SnapshotValue.of(data)
         if not isinstance(data, SnapshotValue):
-            raise TypeError("Data must be either binary, string, or SnapshotValue")
-        return Snapshot(data, {})
+            data = SnapshotValue.of(data)
+        return Snapshot(data, ArrayMap.empty())
 
     @staticmethod
-    def of_entries(entries: Dict[str, SnapshotValue]) -> "Snapshot":
-        subject: Optional[SnapshotValue] = None
-        facet_data: Dict[str, SnapshotValue] = {}
-        for key, value in entries.items():
-            if not key:
-                if subject is not None:
-                    raise ValueError("Duplicate root snapshot detected")
-                subject = value
-            else:
-                facet_data[key] = value
-        return Snapshot(
-            subject if subject is not None else SnapshotValue.of(""), facet_data
-        )
+    def of_entries(entries: Union[ArrayMap[str, SnapshotValue], None]) -> "Snapshot":
```

Comment:
Don't accept `None`.

Reply:
I make changes to of_entries, that much closely align with the kotlin code.